fastparquet/speedups.pyx

```diff
@@ -4,164 +4,149 @@ Native accelerators for Parquet encoding and decoding.
+from cpython cimport (array, PyBytes_GET_SIZE, PyBytes_AS_STRING, PyBytes_Check,
+                      PyBytes_FromStringAndSize, PyUnicode_AsUTF8String)
+from cpython.buffer cimport (PyBUF_ANY_CONTIGUOUS, PyObject_GetBuffer,
+                             PyBuffer_Release)
+from libc.string cimport memcpy
+from libc.stdlib cimport malloc, free
+
+PY2 = sys.version_info[0] == 2
 _obj_dtype = np.dtype('object')
 
+cdef extern from "Python.h":
+    bytearray PyByteArray_FromStringAndSize(char *v, Py_ssize_t l)
+    char* PyByteArray_AS_STRING(object string)
+    object PyUnicode_FromStringAndSize(const char *u, Py_ssize_t size)
+    int PyUnicode_Check(object text)
+    int PyBytes_AsStringAndSize(bytes obj, char **buffer, Py_ssize_t *length)
+    char* PyUnicode_AsUTF8AndSize(unicode unicode, Py_ssize_t *size)
+    unicode PyUnicode_DecodeUTF8(char *s, Py_ssize_t size, char *errors)
 
-def _to_array(obj):
-    """
-    Convert *obj* (a ndarray-compatible object, e.g. pandas Series)
-    to a true ndarray.
-    """
-    if not isinstance(obj, np.ndarray):
-        try:
-            obj = obj.__array__()
-        except AttributeError:
-            raise TypeError("expected an ndarray-compatible object, got %r"
-                            % (type(obj,)))
-        assert isinstance(obj, np.ndarray)
-    return obj
-
-
-def _check_1d_object_array(np.ndarray arr):
-    if arr.ndim != 1:
-        raise TypeError("expected a 1d array")
-    if arr.dtype != _obj_dtype:
-        raise TypeError("expected an object array")
-
-
-def array_encode_utf8(inp):
-    """
-    utf-8 encode all elements of a 1d ndarray of "object" dtype.
-    A new ndarray of bytes objects is returned.
-    """
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def encode(buf, bint utf8=1):
     cdef:
-        Py_ssize_t i, n
-        np.ndarray[object] arr
-        np.ndarray[object] result
+        Py_ssize_t i, l, n_items, data_length, total_length
+        object[:] input_values
+        char** encoded_values
+        int[:] encoded_lengths
+        char* encv
+        char* b
+        bytearray out
+        char* data
+        object u
+
+    # normalise input
+    input_values = np.asanyarray(buf, dtype=object).reshape(-1, order='A')
+
+    # determine number of items
+    n_items = input_values.shape[0]
+
+    # setup intermediates
+    encoded_values = <char **>malloc(n_items * sizeof(char *))
+    encoded_lengths = np.empty(n_items, dtype=np.intc)
+
+    # first iteration to convert to bytes
+    data_length = 0
+    for i in range(n_items):
+        u = input_values[i]
+        if utf8:
+            if not PyUnicode_Check(u):
+                raise TypeError('expected unicode string, found %r' % u)
+            encoded_values[i] = PyUnicode_AsUTF8AndSize(u, &l)
+        else:
+            PyBytes_AsStringAndSize(u, &encv, &l)
+            encoded_values[i] = encv
+        data_length += l + 4  # 4 bytes to store item length
+        encoded_lengths[i] = l
```

Comment:
What if `l` is larger than `2**31`? Should we detect it and raise `OverflowError`?